Last call changes (#183)
securitytxt/security-txt: draft-foudil-securitytxt.md

security researchers, they
often lack the channels to report them properly. As a result,
security vulnerabilities may be left unreported. This document defines a format
("security.txt") to help organizations describe their vulnerability disclosure practices

> in order to make it easier for researchers to report security vulnerabilities.

Reviewer: ```suggestion
to make it easier for researchers to report security vulnerabilities.
```

Author: will change

As per section 4 of {{?RFC2142}}, there is an existing convention
of using the \<SECURITY@domain\> email address for communications regarding
security vulnerabilities. That convention provides only a single, email-based
channel of communication for security vulnerabilities per domain, and does not provide
a way for domain owners to publish information about their security disclosure
practices.

There are also contact conventions prescribed for Internet Service Providers (ISPs)
in section 2 of {{?RFC3013}}, for Computer Security Incident Response Teams (CSIRTs)
in section 3.2 of {{?RFC2350}} and for site operators in section 5.2 of
{{?RFC2196}}. As per {{?RFC7485}}, there is also contact information provided by
Regional Internet Registries (RIRs) and domain registries for owners of IP
addresses, autonomous system numbers (ASNs) and domain names. However, none of
these address the issue of how security researchers can locate vulnerability disclosure
practices and contact information for organizations in order to report
security vulnerabilities.

In this document, we define a richer and extensible way
for organizations to communicate information about their security disclosure

> practices and ways to contact them, which is not limited to email and also allows for additional features

Reviewer: This sentence is too wordy.

Author: Rewording this - being that we already mentioned this is intended for researchers to be consumed, I took that out and only left the part about the organizations.

such as encryption. This format is designed to help
assist with the security disclosure process by making it easier
for organizations to designate the preferred steps for researchers to take
when trying to reach out to them with security vulnerabilities as well as provide

> additional information about their vulnerability disclosure practices useful to security researchers.

Reviewer: This sentence is too wordy. Needs rephrasing.

Author: will do

# The Specification

This document defines a text file to be placed in a known location
that provides information about the vulnerability disclosure practices of a particular organization.
This is intended to help security researchers when disclosing security vulnerabilities.

The file is named "security.txt", and this file MUST be placed under the
/.well-known/ path ("/.well-known/security.txt") {{!RFC8615}} of a domain name or IP address for web
properties. For legacy compatibility, a security.txt file might be placed at the top level path (see {{weblocation}}).

For web-based services, the file MUST be accessible via the Hypertext Transfer Protocol (HTTP) 1.0 {{!RFC1945}} or higher version, as a resource of Internet Media Type "text/plain" with the default charset parameter set to "utf-8" per section 4.1.3 of {{!RFC2046}}, and it MUST be served with "https" (as per section 2.7.2 of {{!RFC7230}}). For file systems a "security.txt" file SHOULD be placed in the root directory of a particular file system.

> This text file contains multiple fields with different values. A field contains a name which is the first part of a field all the way up

Author: will do

that provides information about the vulnerability disclosure practices of a particular organization.
This is intended to help security researchers when disclosing security vulnerabilities.

By convention, the file is named "security.txt".

When made available on HTTP servers, it MUST be placed under the
/.well-known/ path (as "/.well-known/security.txt") {{!RFC8615}} of a domain name or IP address.
For legacy compatibility, a security.txt file might be placed at the top level path (see {{weblocation}}).

> For file systems a "security.txt" file SHOULD be placed in the root directory of a particular file system.

Reviewer: ```suggestion
For file systems a "security.txt" file SHOULD be placed in the root directory of the file system.
```

I do not think "particular" is needed here.

Author: done

> On HTTP servers, the file MUST be accessed via the Hypertext Transfer Protocol (HTTP) 1.0 {{!RFC1945}} or higher version

Reviewer: ```suggestion
On HTTP servers, the file MUST be accessed via the Hypertext Transfer Protocol (HTTP) 1.0 {{!RFC1945}} or a higher version
```

Author: done

> and the "https" scheme (as per section 2.7.2 of {{!RFC7230}}). It MUST have a Content-Type of "text/plain"

Reviewer: > On HTTP servers, the file MUST be accessed via the Hypertext Transfer Protocol 1.0 or a higher version and the "https" scheme.

This sentence is quite difficult to follow. Especially with all the inline references. Might need rephrasing.

Author: reworded

# The Specification

This document defines a text file to be placed in a known location
that provides information for security researchers about the vulnerability disclosure

> practices of a particular organization in order to assist them in disclosing security vulnerabilities.

Reviewer: ```suggestion
practices of a particular organization to assist them in disclosing security vulnerabilities.
```

Author: will do

A "security.txt" file MUST only apply to the domain or IP address in the URI used to retrieve it,
not to any of its subdomains or parent domains. A "security.txt" file that is found
in a file system MUST only apply to the folder
in which it is located, and not to any of its parent or sibling folders.

> However, it will apply to all subfolders.

Reviewer: I would reword this as follows:

> A "security.txt" file that is found in a file system MUST only apply to the folder in which it is located and that folder's subfolders. The file does not apply to any of the folder's parent or sibling folders.

Author: done

Encryption: openpgp4fpr:5f2de5521c63a801ab59ccb603d49de44b29100f
~~~~~~~~~~

### Expires {#expires}

This field indicates the date/time after which the data contained in the "security.txt"
file is considered stale and should not be used.
The value of this field follows the format defined in section 3.3 of {{!RFC5322}}.

This field MUST NOT appear more than once.

~~~~~~~~~~

> Expires: Thu, 31 Dec 2020 18:37:07 -0800

Reviewer: Is the date arbitrary? Is it the implementer's sole discretion to set a realistic expiration date? How does one determine what a suitable expiration date is?

Author: I added a reference to the staleness section from the security considerations section

### Policy {#policy}

> This field indicates a link to where the vulnerability disclosure policy is located.

Reviewer: ```suggestion
This field indicates a link to where the vulnerability-disclosure policy is located.
```

Author: not sure why a dash would work here?

This can help security researchers understand

> the organization's vulnerability reporting practices.

Reviewer: ```suggestion
the organization's vulnerability-reporting practices.
```

Author: same question about the dash as above

## Web-based services {#weblocation}

Web-based services MUST place the security.txt file under the "/.well-known/" path; e.g. https://example.com/.well-known/security.txt
as per {{!RFC8615}}. For legacy compatibility, a security.txt file might be placed at the top-level path

> or redirect (as per section 6.4 of {{!RFC7231}}) to the security.txt file under the /.well-known/ path. If a "security.txt" file

Reviewer: ```suggestion
or redirect (as per section 6.4 of {{!RFC7231}}) to the security.txt file under the "/.well-known/" path. If a "security.txt" file
```

Author: changed

sign-footer      =  < OpenPGP signature from section 7 of [RFC4880] >

unsigned         =  *line (contact-field eol)
                    *line [expires-field eol]
                    *line [lang-field eol] *line
                    ; the order of fields within the file is not important

line             =  (field / comment) eol

eol              =  *WSP [CR] LF

field            =  ack-field /

>                     can-field /

Author: As per IETF feedback, this was changed so Canonical can appear multiple times. This was done to address the use case where multiple redirects point to the same file.

This can result in security reports not being received by the organization
or sent to the attacker.

To protect against this, organizations should digitally sign their "security.txt"
files (as per {{signature}}), use the "Canonical" field to sign the locations
of the file (as per {{canonical}}), and regularly monitor the file and
the referenced resources to detect tampering.

Security researchers should triage the "security.txt" file including verifying
the digital signature and checking any available historical records before using the information

> contained in the file. If "security.txt" file looks suspicious or compromised,

Reviewer 2: (Not introduced by this PR, but probably good to fix this typo). One of--
```suggestion
 contained in the file. If a "security.txt" file looks suspicious or compromised,
```
```suggestion
 contained in the file. If the "security.txt" file looks suspicious or compromised,
```

Author: will do

that permission to test that particular site or project is denied. This might result in pushback against
researchers reporting security issues to that organization.

Therefore, implementors shouldn't assume that presence or absence of
a "security.txt" file grants or denies permission for security testing.

> Any such permission may be indicates in the company's vulnerability disclosure policy

Reviewer: ```suggestion
Any such permission may be indicated in the company's vulnerability-disclosure policy
```

Author: will do

handshake, implementors should validate the provided X.509 certificate
in accordance with {{!RFC6125}} and the following considerations:

- Matching is performed only against the DNS-ID identifiers.

- DNS domain names in server certificates MAY contain the wildcard
character '*' as the complete left-most label within the identifier.

The certificate may also be checked for revocation via the Online Certificate Status
Protocol (OCSP) {{!RFC6960}}, certificate revocation lists (CRLs), or similar mechanisms.

In cases where the "security.txt" file cannot be served via HTTPS or is
being served with an invalid certificate, additional human triage is recommended since
the contents may have been modified while in transit.

As an additional layer of protection, it is also recommended that
organizations digitally sign their "security.txt" file with OpenPGP (as per {{signature}}).
Also, to protect security reports from being tampered with or observed while in transit,
organizations should specify encryption keys (as per {{encryption}}) unless
HTTPS is being used.

However, the determination of validity of such keys is out of scope

> for this specification. Researches need to establish other secure means to

Reviewer: ```suggestion
for this specification. Security researches need to establish other secure means to
```

Author: will change

Reviewer: Actually, there is a typo there. It should be:

```suggestion
for this specification. Security researchers need to establish other secure means to
```

Author: I fixed it

would become easier once a "security.txt" file is published by
an organization. In addition, there is an increased likelihood of reports
being sent in an automated fashion and/or as result of automated scans without
human triage. Attackers can also use this file as a way to spam unrelated
third parties by listing their resources and/or contact information.

Organizations need to weigh the advantages of publishing this file versus
the possible disadvantages and increased resources required to triage
security reports.

Security researchers should consult the organization's vulnerability disclosure policy, if available,
and review the contact information and/or resources referenced within the "security.txt"

> file before submitting reports in an automated fashion or as resulting from automated scans.

Reviewer: Too wordy. Needs rephrasing.

Author: will do
